src/openrc-run/openrc-run.c

```diff
@@ -423,61 +425,61 @@ svc_exec(const char *arg1, const char *arg2)
 		/* UNREACHABLE */
 	}
 
-	buffer = xmalloc(sizeof(char) * BUFSIZ);
 	fd[0].fd = signal_pipe[0];
+	fd[1].fd = master_tty;
 	fd[0].events = fd[1].events = POLLIN;
 	fd[0].revents = fd[1].revents = 0;
-	if (master_tty >= 0) {
-		fd[1].fd = master_tty;
-		fd[1].events = POLLIN;
-		fd[1].revents = 0;
-	}
 
+	next_warn = clock_ms() + (WARN_TIMEOUT * 1000);
 	for (;;) {
-		if ((s = poll(fd, master_tty >= 0 ? 2 : 1, WAIT_INTERVAL / 1000000)) == -1) {
+		int timeout;
+		if (forever) {
+			timeout = -1;
+		} else if ((timeout = next_warn - clock_ms()) < 0) {
+			timeout = 0;
+		}
+
+		if ((s = poll(fd, master_tty >= 0 ? 2 : 1, timeout > 0 ? timeout : 0)) == -1) {
 			if (errno != EINTR) {
 				eerror("%s: poll: %s", applet, strerror(errno));
 				ret = -1;
 				break;
 			}
+			continue;
 		}
 
-		if (s == 0 && !forever) {
-			timespecsub(&timeout, &interval, &timeout);
-			if (timeout.tv_sec <= 0) {
-				kill(service_pid, SIGKILL);
-				ret = -1;
-				break;
-			}
-			timespecsub(&warn, &interval, &warn);
-			if (warn.tv_sec <= 0) {
-				ewarn("%s: waiting for service (%d seconds)", applet, (int)timeout.tv_sec);
-				warn = (struct timespec) { .tv_sec = WARN_TIMEOUT };
-			}
-		} else if (s > 0) {
-			if (fd[1].revents & (POLLIN | POLLHUP)) {
-				bytes = read(master_tty, buffer, BUFSIZ);
-				write_prefix(buffer, bytes, &prefixed);
+		if (s == 0) {
+			if ((kill_timeout -= WARN_TIMEOUT) > 0) {
```

Comment:
Use an absolute target for this also. You can set it right before the `fork()` call. And you can also do `next_warn = min(kill_timeout, next_warn+10)` so that we don't exceed the kill timer.